sql/api/src/main/scala/org/apache/spark/sql/types/DataType.scala

```diff
@@ -408,6 +408,37 @@ object DataType {
     }
   }
 
+  /**
+   * Check if `from` is equal to `to` type except for collations and nullability, which are
+   * both checked to be compatible so that data of type `from` can be interpreted as of type `to`.
+   */
+  private[sql] def equalsIgnoreCompatibleCollationAndNullability(
+      from: DataType,
+      to: DataType): Boolean = {
+    (from, to) match {
+      case (_: StringType, _: StringType) => true
+
+      case (ArrayType(fromElement, fn), ArrayType(toElement, tn)) =>
+        (tn || !fn) && equalsIgnoreCompatibleCollationAndNullability(fromElement, toElement)
```

Comment:
Why do we want to allow setting the type nullable?

Reply:
Changed the approach, we now only check for possible collation difference; nullability must remain the same. In this way by checking only for possible collation difference we scope down the changes whereby previously we were checking for complete data type equality (only comment changes were allowed in alter column command).